### PR TITLE
Test InstallApp and ensureBot

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

Tests for PR https://github.com/mattermost/mattermost-plugin-apps/pull/142

I've introduced some `Client4` mocking in this PR. I don't want it to hold up https://github.com/mattermost/mattermost-plugin-apps/pull/142, so I made this a separate PR.

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-34577
Fixes https://mattermost.atlassian.net/browse/MM-34195

### CODE REVIEW

File: server/clients/client4.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	UpdateUserRoles(userID, newRoles string) (bool, *model.Response)
+	CreateUserAccessToken(userID, name string) (*model.UserAccessToken, *model.Response)
+	GetOAuthApp(clientID string) (*model.OAuthApp, *model.Response)
+	CreateOAuthApp(*model.OAuthApp) (*model.OAuthApp, *model.Response)
+}
+
+/*
+	asAdmin := model.NewAPIv4Client(conf.MattermostSiteURL)
+	asAdmin.SetToken(session.Token)
+*/
```
Comment (hanzei) [inline server/clients/client4.go:20]: Remove?

File: server/clients/client4.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+type ClientService interface {
+	NewClient4(siteURL string) Client4
+}
```
Comment (hanzei) [inline server/clients/client4.go:24]: Is that level on direction really needed? Are additional clients expected to be added?

File: server/proxy/install_test.go [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+	appStore.EXPECT().Get(appID).Times(2).Return(app, nil)
+	appStore.EXPECT().Save(app).Return(nil)
+
+	upstreams := map[apps.AppID]upstream.Upstream{}
+	up := mock_upstream.NewMockUpstream(ctrl)
+	cr := &apps.CallResponse{
+		Markdown: "Install done",
+	}
+	bb, _ := json.Marshal(cr)
```
Comment (hanzei) [inline server/proxy/install_test.go:76]: Why not check the error to double check?

**Review by hanzei — APPROVED** [review]
Urg, the amount of mocking that needs to be done for the test seems very inflexible to me. I don't have an objection to merging this, but it got me thinking how tests can be written without depending on so much.

**Comment by mickmister** [comment]
@levb Do you think we should be adopting tests like this (mocking Client4 methods individually etc)? If not, I think this PR should be closed

**Comment by levb** [comment]
@mickmister 0/5. Seems like a lot of work, and we have the `test-e2e` that allows us to exercise these paths, no?

**Comment by mickmister** [comment]
Closing as it doesn't seem to be providing any value. We can reference this PR if we'd like to add more unit testing to the framework.

**Pull request closed without merging**